src/transformStateWithClones.js

```diff
@@ -7,7 +7,59 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesHistory = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    stateCopy = { ...stateCopy };
+
+    switch (action.type) {
+      case 'addProperties':
+        stateCopy = addProperties(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        stateCopy = removeProperties(stateCopy, action.keysToRemove);
```

Comment:
Similar to the `addProperties` function, `removeProperties` should also modify the `stateCopy` directly without returning a new object.